src/expr/node_view.cpp

```diff
@@ -0,0 +1,118 @@
+/*********************                                                        */
+/*! \file node_view.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andres Noetzli
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Classes that provide different views on nodes
+ **
+ ** The classes in this file provide different views on a given node (e.g. an
+ ** iterator that treats a certain kind as flattened).
+ **/
+
+#include "expr/node_view.h"
+
+namespace CVC4 {
+namespace expr {
+
+template <bool ref_count>
+FlatViewTemplate<ref_count>::FlatViewTemplate(NodeTemplate<ref_count> node,
+                                              Kind kind,
```

Comment:
Hmm, that is a good question, especially given the `Assert(node.getKind() == kind)` below. I think my original intention was that you could say that if the `kind` does not match the node kind, it would be treated as if it was an application of the `kind` with a single argument (which can be useful in some cases, e.g. if your top node is either a single node or a conjunction of nodes and you want to loop over them uniformly). I'll remove.